bluemira/equilibria/harmonics.py

```diff
@@ -0,0 +1,682 @@
+        automatically if relevent).
+    plamsa_bounday:
+        XZ coordinates of the plasma boundary
+    point_type: string
+        Method for creating a set of points: 'arc', 'arc_plus_extrema',
+        'random', or 'random_plus_extrema'
+
+    Returns
+    -------
+    collocation: dict
+        Dictionary containing collocation points:
+        - "x" and "z" values of collocation points.
+        - "r" and "theta" values of collocation points.
+
+    """
+    point_type = PointType[point_type.upper()]
+    x_bdry = plamsa_bounday.x
+    z_bdry = plamsa_bounday.z
+
+    if point_type in (PointType.ARC, PointType.ARC_PLUS_EXTREMA):
+        # Hello spherical coordinates
+        theta_bdry = np.arctan2(x_bdry, z_bdry)
+
+        # Equispaced arc
+        collocation_theta = np.linspace(
+            np.amin(theta_bdry), np.amax(theta_bdry), n_points + 2
+        )
+        collocation_theta = collocation_theta[1:-1]
+        collocation_r = 0.9 * np.amax(x_bdry) * np.ones(n_points)
+
+        # Cartesian coordinates
+        collocation_x = collocation_r * np.sin(collocation_theta)
+        collocation_z = collocation_r * np.cos(collocation_theta)
+
+    if point_type in (PointType.RANDOM, PointType.RANDOM_PLUS_EXTREMA):
+        # Random sample within a circle enclosed by the LCFS
+        half_sample_x_range = 0.5 * (np.max(x_bdry) - np.min(x_bdry))
+        sample_r = half_sample_x_range * np.random.rand(n_points)
+        sample_theta = (np.random.rand(n_points) * 2 * np.pi) - np.pi
+
+        # Cartesian coordinates
+        collocation_x = (
+            sample_r * np.sin(sample_theta) + np.min(x_bdry) + half_sample_x_range
+        )
+        collocation_z = sample_r * np.cos(sample_theta) + z_bdry[np.argmax(x_bdry)]
+
+        # Spherical coordinates
+        collocation_r = np.sqrt(collocation_x**2 + collocation_z**2)
+        collocation_theta = np.arctan2(collocation_x, collocation_z)
+
+    if point_type in (PointType.ARC_PLUS_EXTREMA, PointType.RANDOM_PLUS_EXTREMA):
+        # Extrema
+        d = 0.1
+        extrema_x = np.array(
+            [
+                np.amin(x_bdry) + d,
+                np.amax(x_bdry) - d,
+                x_bdry[np.argmax(z_bdry)],
+                x_bdry[np.argmin(z_bdry)],
+            ]
+        )
+        extrema_z = np.array(
+            [
+                0,
+                0,
+                np.amax(z_bdry) - d,
+                np.amin(z_bdry) + d,
+            ]
+        )
+
+        # Equispaced arc + extrema
+        collocation_x = np.concatenate([collocation_x, extrema_x])
+        collocation_z = np.concatenate([collocation_z, extrema_z])
+
+        # Hello again spherical coordinates
+        collocation_r = np.sqrt(collocation_x**2 + collocation_z**2)
+        collocation_theta = np.arctan2(collocation_x, collocation_z)
+
+    return {
+        "r": collocation_r,
+        "theta": collocation_theta,
+        "x": collocation_x,
+        "z": collocation_z,
+    }
```

Comment:
Can we have a Collocation dataclass, it helps with typing and is easier to use in other bits of the code

```python
@dataclass
class Collocation:
    r: np.ndarray
    theta: np.ndarray
    x: np.ndarray
    z: np.ndarray
```
```suggestion
    return Collocation(collocation_r, collocation_theta, collocation_x, collocation_z)
```